### ISSUE TITLE
Map key checker fails with Map implementations using delegate pattern

### ISSUE THREAD

**Issue by atomicknight** [typetools/checker-framework#513]
Environment: Checker Framework 1.9.8, Oracle JDK 1.8.0_65, Windows 7

As part of my code base, I have an implementation of `java.util.Map` that delegates to an internal `Map` instance. The map key checker does not seem to properly handle this case.

First, without any `@KeyFor` annotations:
```java
public class MapImpl1 implements Map<String, String> {
	private final Map<String, String> delegate = new HashMap<>();

	@Override
	public Set<String> keySet() {
		return delegate.keySet();
	}

	@Override
	public Set<Map.Entry<String, String>> entrySet() {
		return delegate.entrySet();
	}

	// Other methods
}
```
This fails with the following errors:
```
d:\workspace\cf-testcases\mapkey1\src\main\java\org\example\MapImpl1.java:60: error: [override.return.invalid] @UnknownKeyFor Set<@UnknownKeyFor String> keySet(@UnknownKeyFor MapImpl1 this) in org.example.MapImpl1 cannot override @UnknownKeyFor Set<@KeyFor("this") String> keySet(@UnknownKeyFor Map<@UnknownKeyFor String, @UnknownKeyFor String> this) in java.util.Map; attempting to use an incompatible return type
	public Set<String> keySet() {
	          ^
  found   : @UnknownKeyFor Set<@UnknownKeyFor String>
  required: @UnknownKeyFor Set<@KeyFor("this") String>
d:\workspace\cf-testcases\mapkey1\src\main\java\org\example\MapImpl1.java:61: error: [return.type.incompatible] incompatible types in return.
		return delegate.keySet();
		                      ^
  found   : @UnknownKeyFor Set<@KeyFor("this.delegate") String>
  required: @UnknownKeyFor Set<@UnknownKeyFor String>
d:\workspace\cf-testcases\mapkey1\src\main\java\org\example\MapImpl1.java:70: error: [override.return.invalid] @UnknownKeyFor Set<@UnknownKeyFor Entry<@UnknownKeyFor String, @UnknownKeyFor String>> entrySet(@UnknownKeyFor MapImpl1 this) in org.example.MapImpl1 cannot override @UnknownKeyFor Set<@UnknownKeyFor Entry<@KeyFor("this") String, @UnknownKeyFor String>> entrySet(@UnknownKeyFor Map<@UnknownKeyFor String, @UnknownKeyFor String> this) in java.util.Map; attempting to use an incompatible return type
	public Set<Map.Entry<String, String>> entrySet() {
	          ^
  found   : @UnknownKeyFor Set<@UnknownKeyFor Entry<@UnknownKeyFor String, @UnknownKeyFor String>>
  required: @UnknownKeyFor Set<@UnknownKeyFor Entry<@KeyFor("this") String, @UnknownKeyFor String>>
d:\workspace\cf-testcases\mapkey1\src\main\java\org\example\MapImpl1.java:71: error: [return.type.incompatible] incompatible types in return.
		return delegate.entrySet();
		                        ^
  found   : @UnknownKeyFor Set<@UnknownKeyFor Entry<@KeyFor("this.delegate") String, @UnknownKeyFor String>>
  required: @UnknownKeyFor Set<@UnknownKeyFor Entry<@UnknownKeyFor String, @UnknownKeyFor String>>
```

Annotating the parameters with `@KeyFor("this")` yields the following:
```
d:\workspace\cf-testcases\mapkey1\src\main\java\org\example\MapImpl2.java:61: error: [return.type.incompatible] incompatible types in return.
		return delegate.keySet();
		                      ^
  found   : @UnknownKeyFor Set<@KeyFor("this.delegate") String>
  required: @UnknownKeyFor Set<@KeyFor("this") String>
d:\workspace\cf-testcases\mapkey1\src\main\java\org\example\MapImpl2.java:71: error: [return.type.incompatible] incompatible types in return.
		return delegate.entrySet();
		                        ^
  found   : @UnknownKeyFor Set<@UnknownKeyFor Entry<@KeyFor("this.delegate") String, @UnknownKeyFor String>>
  required: @UnknownKeyFor Set<@UnknownKeyFor Entry<@KeyFor("this") String, @UnknownKeyFor String>>
```

Annotating the parameters with `@KeyFor("this.delegate")` yields the inverse:
```
d:\workspace\cf-testcases\mapkey1\src\main\java\org\example\MapImpl3.java:60: error: [override.return.invalid] @UnknownKeyFor Set<@KeyFor("this.delegate") String> keySet(@UnknownKeyFor MapImpl3 this) in org.example.MapImpl3 cannot override @UnknownKeyFor Set<@KeyFor("this") String> keySet(@UnknownKeyFor Map<@UnknownKeyFor String, @UnknownKeyFor String> this) in java.util.Map; attempting to use an incompatible return type
	public Set<@KeyFor("this.delegate") String> keySet() {
	          ^
  found   : @UnknownKeyFor Set<@KeyFor("this.delegate") String>
  required: @UnknownKeyFor Set<@KeyFor("this") String>
d:\workspace\cf-testcases\mapkey1\src\main\java\org\example\MapImpl3.java:70: error: [override.return.invalid] @UnknownKeyFor Set<@UnknownKeyFor Entry<@KeyFor("this.delegate") String, @UnknownKeyFor String>> entrySet(@UnknownKeyFor MapImpl3 this) in org.example.MapImpl3 cannot override @UnknownKeyFor Set<@UnknownKeyFor Entry<@KeyFor("this") String, @UnknownKeyFor String>> entrySet(@UnknownKeyFor Map<@UnknownKeyFor String, @UnknownKeyFor String> this) in java.util.Map; attempting to use an incompatible return type
	public Set<Map.Entry<@KeyFor("this.delegate") String, String>> entrySet() {
	          ^
  found   : @UnknownKeyFor Set<@UnknownKeyFor Entry<@KeyFor("this.delegate") String, @UnknownKeyFor String>>
  required: @UnknownKeyFor Set<@UnknownKeyFor Entry<@KeyFor("this") String, @UnknownKeyFor String>>
```

Finally, annotating with `@KeyFor({"this", "this.delegate"})` yields all the mismatches from the original case:
```
d:\workspace\cf-testcases\mapkey1\src\main\java\org\example\MapImpl4.java:60: error: [override.return.invalid] @UnknownKeyFor Set<@KeyFor({"this", "this.delegate"}) String> keySet(@UnknownKeyFor MapImpl4 this) in org.example.MapImpl4 cannot override @UnknownKeyFor Set<@KeyFor("this") String> keySet(@UnknownKeyFor Map<@UnknownKeyFor String, @UnknownKeyFor String> this) in java.util.Map; attempting to use an incompatible return type
	public Set<@KeyFor({"this", "this.delegate"}) String> keySet() {
	          ^
  found   : @UnknownKeyFor Set<@KeyFor({"this", "this.delegate"}) String>
  required: @UnknownKeyFor Set<@KeyFor("this") String>
d:\workspace\cf-testcases\mapkey1\src\main\java\org\example\MapImpl4.java:61: error: [return.type.incompatible] incompatible types in return.
		return delegate.keySet();
		                      ^
  found   : @UnknownKeyFor Set<@KeyFor("this.delegate") String>
  required: @UnknownKeyFor Set<@KeyFor({"this", "this.delegate"}) String>
d:\workspace\cf-testcases\mapkey1\src\main\java\org\example\MapImpl4.java:70: error: [override.return.invalid] @UnknownKeyFor Set<@UnknownKeyFor Entry<@KeyFor({"this", "this.delegate"}) String, @UnknownKeyFor String>> entrySet(@UnknownKeyFor MapImpl4 this) in org.example.MapImpl4 cannot override @UnknownKeyFor Set<@UnknownKeyFor Entry<@KeyFor("this") String, @UnknownKeyFor String>> entrySet(@UnknownKeyFor Map<@UnknownKeyFor String, @UnknownKeyFor String> this) in java.util.Map; attempting to use an incompatible return type
	public Set<Map.Entry<@KeyFor({"this", "this.delegate"}) String, String>> entrySet() {
	          ^
  found   : @UnknownKeyFor Set<@UnknownKeyFor Entry<@KeyFor({"this", "this.delegate"}) String, @UnknownKeyFor String>>
  required: @UnknownKeyFor Set<@UnknownKeyFor Entry<@KeyFor("this") String, @UnknownKeyFor String>>
d:\workspace\cf-testcases\mapkey1\src\main\java\org\example\MapImpl4.java:71: error: [return.type.incompatible] incompatible types in return.
		return delegate.entrySet();
		                        ^
  found   : @UnknownKeyFor Set<@UnknownKeyFor Entry<@KeyFor("this.delegate") String, @UnknownKeyFor String>>
  required: @UnknownKeyFor Set<@UnknownKeyFor Entry<@KeyFor({"this", "this.delegate"}) String, @UnknownKeyFor String>>
```

The full test case (Maven build - just execute `mvn`): https://github.com/atomicknight/cf-testcases/tree/master/mapkey1

**Comment by smillst** [comment]
This is correct way to annotate key set:
```
public Set<@KeyFor("this") String> keySet() {
```
This is so that the return type is compatible with the return type of the method it is overriding

```
d:\workspace\cf-testcases\mapkey1\src\main\java\org\example\MapImpl2.java:61: error: [return.type.incompatible] incompatible types in return.
        return delegate.keySet();
                              ^
  found   : @UnknownKeyFor Set<@KeyFor("this.delegate") String>
  required: @UnknownKeyFor Set<@KeyFor("this") String>
```

This error is expected because the MapKey Checker cannot determine if an implementation of Map is correct.  For example, MapImpl1 could implement put like so:

```
String put(String key, String value) {
    return value;
}
```
Then MapImpl1.keySet() would not actually return the keyset.  

If you have verified that the implementation of Map is correct you believe that the implementation of Map is correct, you can [suppress this warning](http://types.cs.washington.edu/checker-framework/current/checker-framework-manual.html#suppresswarnings-annotation).

(I'm closing this issue, but please let us know if this explanation isn't clear or if you have more questions.)

**Comment by atomicknight** [comment]
Thanks for the detailed explanation. I'll make the changes in my code.
